public/javascripts/common/SpeedLimit.js

```diff
@@ -0,0 +1,170 @@
+/**
+ * An indicator that displays the speed limit of the current position's nearest road.
+ *
+ * @param {StreetViewPanorama} panorama Panorama object
+ * @param {function} coords Function that returns current longitude and latitude coordinates
+ * @returns {SpeedLimit} SpeedLimit object with updateSpeedLimit function, container, speedLimit object with number and sub (units, e.g. 'mph'),
+ * speedLimitVisible boolean
+ */
+function SpeedLimit(panorama, coords) {
+    let self = this;
+
+    function _init() {
+        self.container = document.getElementById("speed-limit-sign");
+        self.speedLimit = {
+            number: "",
+            sub: ""
+        };
+        self.speedLimitVisible = false
+        updateSpeedLimit();
+
+        // Listen for position changes.
+        panorama.addListener('position_changed', positionChange);
+    }
+
+    function updateSpeedLimit() {
+        self.container.querySelector("#speed-limit").innerText = self.speedLimit.number;
+        self.container.querySelector("#speed-limit-sub").innerText = self.speedLimit.sub;
+        self.container.style.display = self.speedLimitVisible ? "flex" : "none";
+    }
+
+    function findClosestRoad(data, lat, lon) {
+        // Filter to only be roads, and not foot paths/walk ways.
+        const roadHighwayTypes = [
+            "motorway",
+            "trunk",
+            "primary",
+            "secondary",
+            "tertiary",
+            "unclassified",
+            "residential",
+            "motorway_link",
+            "trunk_link",
+            "primary_link",
+            "secondary_link",
+            "tertiary_link",
+            "living_street",
+            "road"
+        ]
+        const roads = data.elements.filter(el => el.type === "way" && el.tags && el.tags.highway && roadHighwayTypes.includes(el.tags.highway));
+
+        let closestRoad = null;
+        let minDistance = Infinity;
+
+        // Go through all the roads and find the closest one, keeping line segments in mind.
+        // Typical point comparison won't always work so it's better to compare to the segment itself.
+        for (const road of roads) {
+            for (let i = 0; i < road.geometry.length - 1; i++) {
+                const point1 = road.geometry[i];
+                const point2 = road.geometry[i + 1];
+
+                const closestPoint = closestPointOnSegment(
```

Comment:
We're already using a geospatial library called [turf.js](https://turfjs.org/docs/6.5.0/api/along), which should be able to do any sort of geospatial computations that you want. So no need to define your own `closestPointOnSegment` funtion, for example!

Reply:
That works out perfectly! I'll switch it to use turf.js and that should make the code a lot more clean and readable.